src/adaptation/icap/ModXact.cc

```diff
@@ -1341,8 +1341,7 @@ void Adaptation::Icap::ModXact::finalizeLogInfo()
     // XXX: This reply (and other ALE members!) may have been needed earlier.
     al.reply = adapted_reply_;
 
-    if (h->rfc931.size())
-        al.cache.rfc931 = h->rfc931.termedBuf();
+    al.tcpClient = h->clientConnection;
```

Comment:
We need to decide, in ICAP context, whether ALE fields (outside of ALE::icap) represent

* the current ICAP transaction
* the corresponding master transaction
* the corresponding HTTP client-to-Squid transaction
* the corresponding HTTP Squid-to-peer transaction.

Commit bd59d61 says that `http::>h` and `http::<h` fields are for HTTP messages embedded in ICAP request and response. This does does not directly answer the question for other non-`icap` ALE fields, but strongly implies that other non-`icap` fields are meant for the ICAP transaction, not client-Squid or Squid-peer HTTP transactions or the master transaction.

Official code is inconsistent; for example, Adaptation::Icap::ModXact::finalizeLogInfo():

* sets al.cache.caddr to `virgin_request_->client_addr` (i.e. HTTP client address)
* sets al.cache.code to `h->logType` (i.e. ICAP transaction outcome?)
* calls HTTP-only (or, more precisely, ICAP-unaware) prepareLogWithRequestDetails() with adapted_request_ that probably supplies some ICAP transaction statistics to not-ICAP-specific ALE fields

I would like us to answer the primary question the following way. Long term:

* ALE should have no master transaction fields. Those fields belong to MasterXaction. One MasterXaction may correspond to many individual transaction ALEs (e.g., one HTTP client-Squid ALE, one ICAP REQMOD ALE, and one eCAP RESPMOD ALE).
* In ICAP context, ALE is for the ICAP transaction (REQMOD, RESPMOD, or OPTIONS).
* ALE::tcpClient is the primary connection of the corresponding transaction. For HTTP client-Squid transaction, that is the connection accepted by Squid. For HTTP Squid-peer transactions (that do not have a dedicated ALE today but probably should have it!), that is the connection opened by Squid. For ICAP transactions, that is the connection to the ICAP service opened by Squid.
* ALE::cache.rfc931 is the identity of the user that has connected to Squid -- a MasterXaction-level info. In the future, it will not be stored in ALE::cache but will be extracted from via client-Squid transaction ALE, where it will be found in ALE::tcpClient::rfc931. Until then, it will continue to live in ALE because MasterXaction does not have access to client-Squid transaction ALE and ACL-checking code does not have access to MasterXaction (all are complex out-of-scope TODOs).

If the above answer is the best we can give, then I suggest:

* Removing ALE::cache.rfc931
* Adding ALE::acceptedClientConnection as a temporary storage for MasterXaction-level info that future code will extract from ALE::tcpClient of a client-Squid transaction.
* Adding Adaptation::Icap::History::acceptedClientAle to remember the client-Squid transaction ALE in ICAP transactions.
* Setting ALE::acceptedClientConnection to `h->httpClientAle->acceptedClientConnection` in this ICAP code.
* Documenting this answer in PR description.

Reply:
Ok, let's follow your plan.

> Adding ALE::acceptedClientConnection as a temporary storage 

Do you suggest creating a new type for this field or reusing Comm::Connection? In the second case, ALE::acceptedClientConnection and ALE::tcpClient will be equal for the client-Squid ALE, right?

> In ICAP context, ALE is for the ICAP transaction

Does it mean that all ALE fields should represent ICAP transaction? E.g., should ALE::url and ALE::icap::reqUri be the same in this case?

Reply:
> > Adding ALE::acceptedClientConnection as a temporary storage
> 
> Do you suggest creating a new type for this field or reusing Comm::Connection? In the second case, ALE::acceptedClientConnection and ALE::tcpClient will be equal for the client-Squid ALE, right?

I suggest using Comm::Connection for both. During the transition, either the two fields will be equal for client-Squid ALE OR ALE code will consult both fields to extract the right information ("the primary connection of this protocol transaction" vs. "client-Squid connection of the master transaction"). The latter feels like a better option to me, but I cannot be sure without trying to implement it.

> > In ICAP context, ALE is for the ICAP transaction

> Does it mean that all ALE fields should represent ICAP transaction? E.g., should ALE::url and ALE::icap::reqUri be the same in this case?

Eventually, yes, all ALE fields will represent the ICAP transaction and there will be no ALE::icap (but we may get protocol-specific Http::ALE and Adaptation::Icap::ALE classes!). During the transition (e.g., in this PR), we should make those decisions one (affected) field at a time, like we are doing with ALE::acceptedClientConnection above. In this PR, we do not want to start working on upgrading ALE fields that we do not have to touch.